apps/io_tester/io_tester.cc

```diff
@@ -657,6 +665,96 @@ class write_io_class_data : public io_class_data {
+        if (type == directory_entry_type::directory) {
+            return do_start_on_directory(path);
+        }
+        throw std::runtime_error(format("Unsupported storage. {} should be directory", path));
+    }
+
+    future<size_t> issue_request(char *buf, io_intent* intent) override {
+        if (all_files_removed()) {
+            return make_ready_future<size_t>(1u);
+        }
+
+        const auto fname = get_filename(_file_id_to_remove);
+        ++_file_id_to_remove;
+
+        return remove_file(fname).then([]{
+            return make_ready_future<size_t>(1u);
+        });
+    }
+
+    void emit_results(YAML::Emitter& out) override {
+        out << YAML::Key << "average latency" << YAML::Value << average_latency();
+    }
+
+private:
+    uint64_t files_count() const {
+        return *_config.files_count;
+    }
+
+    bool all_files_removed() const {
+        return files_count() <= _file_id_to_remove;
+    }
+
+    sstring get_filename(uint64_t file_id) const {
+        return format("{}/test-{}-shard-{:d}-file-{}", _dir_path, name(), this_shard_id(), file_id);
+    }
+
+    future<> do_start_on_directory(sstring path) {
+        _dir_path = std::move(path);
+
+        return seastar::parallel_for_each(boost::irange(UINT64_C(0), files_count()), [this] (uint64_t file_id) {
+            const auto fname = get_filename(file_id);
+            const auto fsize = _config.file_size;
+            const auto flags = open_flags::rw | open_flags::create;
+
+            file_open_options options;
+            options.extent_allocation_size_hint = fsize;
+            options.append_is_unlikely = true;
+
+            return open_file_dma(fname, flags, options).then([fname, fsize] (auto f) {
+                return maybe_remove_file(fname).then([f, fsize] () mutable {
+                    return f.size().then([f, fsize] (uint64_t pre_truncate_size) mutable {
+                        return f.truncate(fsize).then([f, fsize, pre_truncate_size] () {
+                            if (pre_truncate_size >= fsize) {
+                                return make_ready_future<>();
+                            }
+
+                            const uint64_t buffer_size{256ul << 10};
+                            const uint64_t buffers_count{static_cast<uint64_t>(fsize / buffer_size) + 1u};
+
+                            return do_with(boost::irange(UINT64_C(0), buffers_count), [f, buffer_size] (auto& buffers_range) mutable {
+                                return max_concurrent_for_each(buffers_range.begin(), buffers_range.end(), 64, [f, buffer_size] (auto buffer_id) mutable {
+                                    auto source_buffer = allocate_and_fill_buffer(buffer_size);
+                                    auto write_position = buffer_id * buffer_size;
+
+                                    return f.dma_write(write_position, source_buffer.get(), buffer_size).discard_result();
```

Comment:
Why it _this_ here? The job is supposed to just unlink files, not unlink and then write something to them

Reply:
`maybe_remove_file()` is used to preserve files from previous run if they exist. Your comment is related to code that is executed during startup - it only prepares files and fills them. So, its logic is as follows:

1. Create N files via parallel_for:
1.1. If file exists and the user specified `preserve_files=true`, then ensure that its size is sufficient.
1.2. If the file exists, but user specified `preserve_files=false` then remove that file and create new one.
1.3. Ensure that either preserved or created file has the desired size and fill it with dummy data.

As far as I understand main(), it is not a part of evaluation. During evaluation only unlink is called.

```
    future<size_t> issue_request(char *buf, io_intent* intent) override {
        if (all_files_removed()) {
            return make_ready_future<size_t>(1u);
        }

        const auto fname = get_filename(_file_id_to_remove);
        ++_file_id_to_remove;

        return remove_file(fname).then([]{
            return make_ready_future<size_t>(1u);
        });
    }
```

Reply:
Well, I feel that I mistaken the semantic of  `maybe_remove_file()` in `io_class_data` - I tried to follow the same approach. The creation of files here will be adjusted.

Reply:
The PR has been adjusted.

This is a part of `do_start_on_directory()` that is called via `start()`. It creates and fills files (if needed), that will be unlinked during the evaluation.